--- a/backport-include/linux/etherdevice.h
+++ b/backport-include/linux/etherdevice.h
@@ -39,7 +39,7 @@ static inline void u64_to_ether_addr(u64
 }
 #endif /* LINUX_VERSION_IS_LESS(4,11,0) */
 
-#if LINUX_VERSION_IS_LESS(5,15,0)
+#if LINUX_VERSION_IS_LESS(5,10,188)
 /**
  * eth_hw_addr_set - Assign Ethernet address to a net_device
  * @dev: pointer to net_device structure
@@ -51,7 +51,7 @@ static inline void eth_hw_addr_set(struc
 {
 	ether_addr_copy(dev->dev_addr, addr);
 }
-#endif /* LINUX_VERSION_IS_LESS(5,15,0) */
+#endif /* LINUX_VERSION_IS_LESS(5,10,188) */
 
 #if LINUX_VERSION_IS_LESS(5,16,0)
 static inline int backport_device_get_mac_address(struct device *dev, char *addr)
